#1 Sprint 1

Galyk-Ilya opened the pull request:
Сконфигурированно Spring приложение
Разработано 2 REST-контроллера, /films & /users
Проработана валидация согласно ТЗ

xccazz committed:
Сконфигурированно Spring приложение
Разработано 2 REST-контроллера, /films & /users
Проработана валидация согласно ТЗ

xccazz committed:
Изменены модификаторы доступа вспомогательных полей и методов в Unit-тестах

kalitva commented on pom.xml:
```diff
@@ -0,0 +1,69 @@
+            <artifactId>spring-boot-starter-web</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>junit</groupId>
```
junit зависимости уже есть в spring-boot-stater-test, удали

kalitva commented on pom.xml:
```diff
@@ -0,0 +1,69 @@
+        <groupId>org.springframework.boot</groupId>
+        <artifactId>spring-boot-starter-parent</artifactId>
+        <version>2.7.8</version>
+        <relativePath/> <!-- lookup parent from repository -->
```
Удали коммент

kalitva commented on pom.xml:
```diff
@@ -0,0 +1,69 @@
+        </dependency>
+        <dependency>
+            <groupId>com.google.code.gson</groupId>
+            <artifactId>gson</artifactId>
```
gson нам не пригодится

kalitva commented on pom.xml:
```diff
@@ -0,0 +1,69 @@
+        </dependency>
+        <dependency>
+            <groupId>com.fasterxml.jackson.core</groupId>
+            <artifactId>jackson-databind</artifactId>
```
jackson уже есть в spring-boot-starter-web

kalitva commented on src/main/java/ru/yandex/practicum/filmorate/FilmorateApplication.java:
```diff
@@ -0,0 +1,11 @@
+
+@SpringBootApplication
+public class FilmorateApplication {
+	public static void main(String[] args) {
```
Замени табы на пробелы. В intellij нажми Edit --> Convert Indents --> To spaces

kalitva commented on src/main/java/ru/yandex/practicum/filmorate/controllers/Controller.java:
```diff
@@ -0,0 +1,31 @@
+public abstract class Controller<T extends Model> {
+
+    protected final HashMap<Integer, T> dataMap = new HashMap<>();
+    Integer counter = 0;
```
private int

kalitva commented on src/main/java/ru/yandex/practicum/filmorate/controllers/Controller.java:
```diff
@@ -0,0 +1,31 @@
+
+    public abstract void validate(T data, boolean isCreate);
+
+
```
Удали лишние пустые строки

kalitva commented on src/main/java/ru/yandex/practicum/filmorate/controllers/Controller.java:
```diff
@@ -0,0 +1,31 @@
+package ru.yandex.practicum.filmorate.controllers;
+
+import ru.yandex.practicum.filmorate.model.Model;
+import java.util.*;
```
Не используй wildcard импорты.

https://google.github.io/styleguide/javaguide.html#s3.3.1-wildcard-imports

Это может служить причиной лишних конфликтов при мерже веток. Настрой intellij чтобы она не схлопывала импорты

https://blog.marcnuri.com/intellij-idea-how-to-disable-wildcard-imports

kalitva commented on src/main/java/ru/yandex/practicum/filmorate/controllers/Controller.java:
```diff
@@ -0,0 +1,31 @@
+        return dataMap.values();
+    }
+
+    public abstract void validate(T data, boolean isCreate);
```
Не используй флаги в параметрах методов или конструкторов, вообще никогда не используй

kalitva commented on src/main/java/ru/yandex/practicum/filmorate/controllers/FilmController.java:
```diff
@@ -0,0 +1,79 @@
+
+    @Override
+    @GetMapping
+    public Collection<Film> getAll() {
```
Не надо переопределять этот метод, просто поставь `@GetMapping` над super реализацией (так как определен префикс в  `RequestMapping`, вызовы будут поступать в нужный контроллер)

То же самое для двух методов ниже

kalitva commented on src/main/java/ru/yandex/practicum/filmorate/controllers/FilmController.java:
```diff
@@ -0,0 +1,79 @@
+    }
+
+    public void validate(Film film, boolean isCreate) {
+        if (isCreate) {
```
При создании просто засетишь новое айди, не надо это проверять

kalitva commented on src/main/java/ru/yandex/practicum/filmorate/controllers/FilmController.java:
```diff
@@ -0,0 +1,79 @@
+            }
+            log.debug("film " + film.getName() + " id" + film.getId() + " изменен");
+        } else {
+            if (!dataMap.containsKey(film.getId())) {
```
Перенеси в `Controller#update`

kalitva commented on src/main/java/ru/yandex/practicum/filmorate/controllers/FilmController.java:
```diff
@@ -0,0 +1,79 @@
+    public void validate(Film film, boolean isCreate) {
+        if (isCreate) {
+            if (dataMap.containsKey(film.getId())) {
+                throw new InputDataException("Невозможно добавить " + film.getName() + ", так как он был добавлен ранее");
```
Сообщение в исключение - только по английски

kalitva commented on src/main/java/ru/yandex/practicum/filmorate/controllers/FilmController.java:
```diff
@@ -0,0 +1,79 @@
+            if (!dataMap.containsKey(film.getId())) {
+                throw new InputDataException("Невозможно обновить данные фильма, " + film.getName() + " не был добавлен ранее");
+            }
+            log.debug("film " + film.getName() + " id" + film.getId() + " изменен");
```
Логирование - по английски

kalitva commented on src/main/java/ru/yandex/practicum/filmorate/controllers/FilmController.java:
```diff
@@ -0,0 +1,79 @@
+        log.debug(film.getName() + " прошел проверки всех полей");
+    }
+
+    public Map<Integer, Film> getFilmMap() {
```
Зачем здесь этот метод? Удали, в контроллере не должно быть публичных методов без маппинга

kalitva commented on src/main/java/ru/yandex/practicum/filmorate/controllers/UserController.java:
```diff
@@ -0,0 +1,66 @@
+        log.debug("update user");
+        return super.update(user);
+    }
+    public void validate(User user, boolean isCreate) {
```
Пожалуйста, пустая линия между объявлениями методов

kalitva commented on src/main/java/ru/yandex/practicum/filmorate/controllers/UserController.java:
```diff
@@ -0,0 +1,66 @@
+    public Map<Integer, User> getUserMap() {
+        return dataMap;
+    }
+}
```
Все те же замечания, что и для `FilmController`

kalitva commented on src/main/java/ru/yandex/practicum/filmorate/exceptions/GlobalException.java:
```diff
@@ -0,0 +1,17 @@
+
+@ControllerAdvice
+public class GlobalException {
+//    @ExceptionHandler
```
Удали закомментированный код, если вдруг понадобится, достанешь из vcs

kalitva commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -0,0 +1,34 @@
+    private void parseLocalDate(String birthdayString) {
+        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd");
+        this.releaseDate = LocalDate.parse(birthdayString, formatter);
+        //Долго не мог обойти ошибку json-a на вход LocalDate поля, так и не нашел решение =\
```
Попробуй так

```
    @JsonFormat(pattern = "yyyy-MM-dd")
    private LocalDate releaseDate;
```

kalitva commented on src/test/java/ru/yandex/practicum/filmorate/FilmorateApplicationTests.java:
```diff
@@ -0,0 +1,13 @@
+package ru.yandex.practicum.filmorate;
```
Давай удалим этот файл

kalitva commented on src/test/java/ru/yandex/practicum/filmorate/controllers/FilmControllerTest.java:
```diff
@@ -0,0 +1,105 @@
+
+    @BeforeEach
+    public void createForTests() {
+        String description200 = "11111111111111111111111111111111111111111111111111111111111111111111111111111111111" +
```
`String description200 = "1".repeat(200)`

kalitva commented on src/main/java/ru/yandex/practicum/filmorate/exceptions/ErrorObject.java:
```diff
@@ -0,0 +1,11 @@
+
+@Getter
+@Setter
+public class ErrorObject {
```
`ErrorObject --> ErrorResponse`

xccazz committed:
Удалили лишние методы, файлы, проверки;
Логирование, сообщения исключений - на английском;
Исключили флаг из validate, wildcard импорты;

xccazz committed:
Убрали неиспользуемый конструктор у моделей

xccazz committed:
Добавили конструкторы у моделей для прохождения тестов gitGub

xccazz committed:
Перенесли поле name из Model в классы наследники

xccazz committed:
--

xccazz committed:
--

xccazz committed:
--

xccazz committed:
--

xccazz committed:
--

xccazz committed:
--

xccazz committed:
9.2

Galyk-Ilya commented:
Долго решал проблему с конфликтом тестов в Idea и Git-а
При аннотировании модели аннотациями: @NoArgsConstructor(force = true), @AllArgsConstructor
создавая объекты, для Unit-тестов, через new, недоступно поле родителя (id) в параметрах конструктора создаваемого наследника (нашел решение правками каждого объекта через сеттер). Но при этом, создавая явный избыточный конструктор в дополнении к созданным аннотацией, GitHub-тесты просто не видели этот конструктор, почему?

kalitva commented on src/main/java/ru/yandex/practicum/filmorate/controllers/Controller.java:
```diff
@@ -0,0 +1,44 @@
+@Slf4j
+public abstract class Controller<T extends Model> {
+
+    protected final HashMap<Integer, T> dataMap = new HashMap<>();
```
`dataMap --> entities`

Избегай в названиях коллекций (и методов соответственно) слова типа `List Map Set Array` и.т.п

kalitva commented on src/main/java/ru/yandex/practicum/filmorate/controllers/Controller.java:
```diff
@@ -0,0 +1,44 @@
+    private int counter = 0;
+
+    @PostMapping
+    public T create(T data) {
```
Лучше `data --> entity`

kalitva commented on src/main/java/ru/yandex/practicum/filmorate/controllers/FilmController.java:
```diff
@@ -0,0 +1,56 @@
+public class FilmController extends Controller<Film> {
+    private final LocalDate theBirthOfCinema = LocalDate.of(1895, 12, 28);
+
+    public Collection<Film> getAll() {
```
Вообще не надо переопределять, также удали `create` и `update`, здесь должен остаться только `validate`. Логирование можешь перенести в родителя

kalitva commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -0,0 +1,27 @@
+public class Film extends Model {
+
+    private String description;
+    protected String name;
```
А почему protected?

kalitva commented on src/main/java/ru/yandex/practicum/filmorate/controllers/Controller.java:
```diff
@@ -0,0 +1,44 @@
+        return dataMap.values();
+    }
+
+    public abstract void validate(T data);
```
Думаю, лучше сделать метод не абстрактным, а просто пустым с возможность переопределить. Могут появиться новые контроллеры, где валидация не требуется

kalitva commented on src/main/java/ru/yandex/practicum/filmorate/model/User.java:
```diff
@@ -0,0 +1,27 @@
+@EqualsAndHashCode
+@NoArgsConstructor(force = true)
+@AllArgsConstructor
+@ToString
```
Если нужно, чтобы айдишник тоже был в `toString` - `@ToString(callSuper = true)`

kalitva commented on src/main/java/ru/yandex/practicum/filmorate/model/User.java:
```diff
@@ -0,0 +1,27 @@
+
+@Getter
+@Setter
+@EqualsAndHashCode
```
`callSuper = true`, если нужно чтобы айдишник участвовал в equals и hashCode

xccazz committed:
9.3

xccazz committed:
9.3 update validation

xccazz committed:
9.3 fix tests

kalitva commented:
> Долго решал проблему с конфликтом тестов в Idea и Git-а При аннотировании модели аннотациями: @NoArgsConstructor(force = true), @AllArgsConstructor создавая объекты, для Unit-тестов, через new, недоступно поле родителя (id) в параметрах конструктора создаваемого наследника (нашел решение правками каждого объекта через сеттер). Но при этом, создавая явный избыточный конструктор в дополнении к созданным аннотацией, GitHub-тесты просто не видели этот конструктор, почему?

Не знаю в чем дело, ломбок и наследование это всегда геморно. Я бы вообще отказался от любых аннотаций для конструкторов. Зачем они нужны? В продакшен коде конструкторы не используется, для тестов можно создание объекта вынести в метод, а в методе сеттерами наполнить объект

kalitva commented on src/main/java/ru/yandex/practicum/filmorate/controllers/Controller.java:
```diff
@@ -0,0 +1,49 @@
+        return entities.values();
+    }
+
+    public void validate(T entity){}
```
protected

kalitva commented on src/main/java/ru/yandex/practicum/filmorate/controllers/FilmController.java:
```diff
@@ -0,0 +1,26 @@
+@RestController
+@RequestMapping("/films")
+public class FilmController extends Controller<Film> {
+    private final LocalDate theBirthOfCinema = LocalDate.of(1895, 12, 28);
```
Оформи константу по конвенции - статическое поле, название в upper case

`private static final LocalDate CINEMA_BIRTHDAY = LocalDate.of(1895, 12, 28);`

https://google.github.io/styleguide/javaguide.html#s5.2.4-constant-names

kalitva commented on src/main/java/ru/yandex/practicum/filmorate/controllers/FilmController.java:
```diff
@@ -0,0 +1,26 @@
+            throw new InputDataException("The release date must be - no earlier than December 28, 1895");
+        }
+        log.debug(film.getName() + " passed the test for releaseDate");
+        log.debug(film.getName() + " passed the checks of all fields");
```
Думаю, это логирование можно убрать

kalitva commented on src/main/java/ru/yandex/practicum/filmorate/controllers/UserController.java:
```diff
@@ -0,0 +1,27 @@
+
+    @Override
+    public void validate(User user) {
+        if (user.getBirthday().isAfter(LocalDate.now())) {
```
Можно воспользоваться аннотацией `@javax.validation.constraints.Past`

kalitva commented on src/main/java/ru/yandex/practicum/filmorate/controllers/UserController.java:
```diff
@@ -0,0 +1,27 @@
+            throw new InputDataException("Date of birth cannot be in the future");
+        }
+        log.debug("user passed the test for Birthday");
+        if (user.getName() == null || user.getName().isBlank()) {
```
Все-таки это не совсем валидация. Я бы предложил вынести в другой метод, назвать например `preSave`. Должно также работать как и с `validate` - пустой метод определяется в родителе, там же вызывается перед сохранением/обновлением. В этом контроллере нужно будет переопределить

kalitva commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -0,0 +1,32 @@
+public class Film extends Model {
+    @Size(min = 1, max = 200, message = "The maximum description length is 200 characters")
+    private String description;
+    @NotBlank(message = "Movie title cannot be empty")
```
Давай отделим пустыми линиями поля с аннотациями

kalitva commented on src/main/java/ru/yandex/practicum/filmorate/model/Model.java:
```diff
@@ -0,0 +1,15 @@
+@Setter
+@ToString
+public abstract class Model {
+    @NotNull
```
Примитив не может быть null

kalitva commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -0,0 +1,32 @@
+    private String name;
+    @JsonFormat(pattern = "yyyy-MM-dd")
+    private LocalDate releaseDate;
+    @Min(value = 1, message = "Movie duration must be positive")
```
Лучше - `@Positive`

kalitva commented on src/test/java/ru/yandex/practicum/filmorate/controllers/FilmControllerTest.java:
```diff
@@ -0,0 +1,99 @@
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+public class FilmControllerTest {
```
Если интересно как правильно тестировать спринговые контроллеры, напиши в чате, отправлю пример

xccazz committed:
9.4
